lightstep/trace_context.py

```diff
@@ -0,0 +1,311 @@
+
+# https://www.w3.org/TR/trace-context/#a-traceparent-is-received
+_HEXDIGITLOWER = r"[abcdef\d]"
+
+# https://www.w3.org/TR/2019/CR-trace-context-20190813/#version
+# https://www.w3.org/TR/2019/CR-trace-context-20190813/#version-format
+_VERSION = re_compile(r"\s*(?P<version>{0}{{2}})-".format(_HEXDIGITLOWER))
+
+# https://www.w3.org/TR/trace-context/#trace-id
+# https://www.w3.org/TR/trace-context/#parent-id
+# https://www.w3.org/TR/trace-context/#trace-flags
+_REMAINDER_MATCH_RE = (
+    r"{0}{{2}}-"
+    r"(?P<trace_id>{0}{{32}})-"
+    r"(?P<parent_id>{0}{{16}})-"
+    r"(?P<trace_flags>{0}{{2}})"
+).format(_HEXDIGITLOWER)
+
+# https://www.w3.org/TR/2019/CR-trace-context-20190813/#versioning-of-traceparent
+_00_VERSION_REMAINDER = re_compile("".join([_REMAINDER_MATCH_RE, r"\s*$"]))
+_FUTURE_VERSION_REMAINDER = re_compile(
+    "".join([_REMAINDER_MATCH_RE, r"(\s*$|-[^\s]+)"])
+)
+
+_KEY_VALUE = re_compile(
+    r"^\s*({0})=({1})\s*$".format(
+        # https://www.w3.org/TR/2019/CR-trace-context-20190813/#tracestate-header
+        # https://www.w3.org/TR/2019/CR-trace-context-20190813/#key
+        (
+            r"[{0}][-{0}\d_*/]{{0,255}}|"
+            r"[{0}\d][-{0}\d_*/]{{0,240}}@[{0}][-{0}\d_*/]{{0,13}}"
+        ).format("a-z"),
+        # https://www.w3.org/TR/2019/CR-trace-context-20190813/#value
+        r"[ {0}]{{0,255}}[{0}]".format(
+            escape(
+                r"".join(
+                    [
+                        chr(character) for character in range(33, 127)
+                        if character not in [44, 61]
+                    ]
+                )
+            )
+        )
+    )
+)
+
+_BLANK = re_compile(r"^\s*$")
+
+
+class TraceContextPropagator(Propagator):
+    """
+    Propagator for the W3C Trace Context format
+
+    The latest version of this Candidate Recommendation can be found here:
+    https://www.w3.org/TR/trace-context/
+
+    This is an implementation for this specific version:
+    https://www.w3.org/TR/2019/CR-trace-context-20190813/
+
+    The Candidate Recommendation will be referred to as "the document" in the
+    comments of this implementation. This implementation adds comments which
+    are URLs that point to the specific part of the document that explain the
+    rationale of the code that follows the comment.
+
+    There is a set of test cases defined for testing any implementation of the
+    document. This set of test cases can be found here:
+    https://github.com/w3c/trace-context/tree/master/test
+
+    This set of test cases will be referred to as "the test suite" in the
+    comments of this implementation.
+    """
+
+    def inject(self, span_context, carrier):
+
+        carrier[_TRACEPARENT] = "00-{}-{}-{}".format(
+            format(span_context.trace_id, "032x"),
+            format(span_context.span_id, "016x"),
+            format(span_context.baggage.pop("trace-flags", 0), "02x")
+        )
+
+        carrier.update(span_context.baggage)
```

Comment:
We are dropping our own baggage here, right? I'm afraid that this will clash with some other fields/data already existing in the carrier. I don't remember this from the W3C, but will double check ;)

Reply:
Do you want this removed now, or should I wait for your double-check?

Reply:
Lets go ahead and keep it like this (open a ticket so we can double verify later ;) )